cmd/builder/internal/builder/main_test.go

```diff
@@ -272,6 +276,101 @@ func TestGenerateAndCompile(t *testing.T) {
 
+// An incomplete set of replace statements in these tests
+// may cause them to fail during the release process, when
+// the local version of modules in the release branch is
+// not yet available on the Go package repository.
+// Unless the replace statements route all modules to the
+// local copy, `go get` will try to fetch the unreleased
+// version remotely and some tests will fail.
+func TestReplaceStatementsAreComplete(t *testing.T) {
+	var err error
+	dir := t.TempDir()
+	cfg := NewDefaultConfig()
+	cfg.Distribution.Go = "go"
+	cfg.Distribution.OutputPath = dir
+	// Use a deliberately nonexistent version to simulate an unreleased
+	// version of the package. Not strictly necessary since this test
+	// will catch gaps in the replace statements before a release is in
+	// progress.
+	cfg.Distribution.OtelColVersion = "1.9999.9999"
+	cfg.Replaces = append(cfg.Replaces, generateReplaces()...)
+	cfg.Providers = &[]Module{}
+	cfg.Exporters, err = parseModules([]Module{
+		{
+			GoMod: "go.opentelemetry.io/collector/exporter/debugexporter v1.9999.9999",
+		},
+		{
+			GoMod: "go.opentelemetry.io/collector/exporter/nopexporter v1.9999.9999",
+		},
+		{
+			GoMod: "go.opentelemetry.io/collector/exporter/otlpexporter v1.9999.9999",
+		},
+		{
+			GoMod: "go.opentelemetry.io/collector/exporter/otlphttpexporter v1.9999.9999",
+		},
+	})
+	require.NoError(t, err)
+	cfg.Receivers, err = parseModules([]Module{
+		{
+			GoMod: "go.opentelemetry.io/collector/receiver/nopreceiver v1.9999.9999",
+		},
+		{
+			GoMod: "go.opentelemetry.io/collector/receiver/otlpreceiver v1.9999.9999",
+		},
+	})
+	require.NoError(t, err)
+	cfg.Extensions, err = parseModules([]Module{
+		{
+			GoMod: "go.opentelemetry.io/collector/extension/zpagesextension v1.9999.9999",
+		},
+	})
+	require.NoError(t, err)
+	cfg.Processors, err = parseModules([]Module{
+		{
+			GoMod: "go.opentelemetry.io/collector/processor/batchprocessor v1.9999.9999",
+		},
+		{
+			GoMod: "go.opentelemetry.io/collector/processor/memorylimiterprocessor v1.9999.9999",
+		},
+	})
+	require.NoError(t, err)
+
+	require.NoError(t, cfg.SetBackwardsCompatibility())
+	require.NoError(t, cfg.Validate())
+	require.NoError(t, cfg.ParseModules())
+	err = GenerateAndCompile(cfg)
+	require.NoError(t, err)
+
+	gomodname := path.Join(dir, "go.mod")
+	gomod, err := os.ReadFile(gomodname)
+	require.NoError(t, err)
+
+	mod, err := modfile.Parse(path.Join(dir, "go.mod"), gomod, nil)
+	require.NoError(t, err)
+
+	replaceMods := map[string]bool{}
+
+	for _, suffix := range replaceModules() {
+		replaceMods[modulePrefix+suffix] = false
+	}
+
+	for _, req := range mod.Require {
+		if !strings.HasPrefix(req.Mod.Path, modulePrefix) {
+			continue
+		}
+
+		_, ok := replaceMods[req.Mod.Path]
+		require.True(t, ok, fmt.Sprintf("Module missing from replace statements list: %s", req.Mod.Path))
+
+		replaceMods[req.Mod.Path] = true
+	}
+
+	for k, v := range replaceMods {
+		require.True(t, v, fmt.Sprintf("Module not used: %s", k))
```

Comment:
This should technically be assert instead of require, but it's OK to leave it like this. Also, you can use Truef, which accepts message and args like this: 

```suggestion
		require.Truef(t, v, "Module not used: %s", k)
```

Reply:
Thanks for the tip on `Truef`, I knew there was something that did that but couldn't remember how it worked.

>This should technically be assert instead of require, but it's OK to leave it like this.
I would actually agree with this for most of our test case assertions, but in this case I think `require` makes more sense. Not stopping test execution after the assertion fails would notify us of all extra modules instead of just the first one that's missing. What do you think?